oarepo_oidc_einfra/tasks.py

```diff
@@ -161,42 +213,140 @@ def update_from_perun_dump(dump_url, fix_communities_in_perun=True):
+    """Synchronize users from perun dump to the database.
+
+    :param dump:                 perun dump data
+    :param community_support:    community support object
+    """
     local_users_by_einfra = einfra_to_local_users_map()
-    for aai_user in dump.users():
-        local_user_id = local_users_by_einfra.pop(aai_user.einfra_id, None)
-        # do not create new users proactively, we can do it on the first login
-        if not local_user_id:
+    for aai_user_chunk in chunks(dump.users(), 100):
+        aai_user_chunk_by_einfra_id = {u.einfra_id: u for u in aai_user_chunk}
+
+        local_user_id_to_einfra_id = {}
+        for einfra_id in aai_user_chunk_by_einfra_id:
+            local_user_id = local_users_by_einfra.pop(einfra_id, None)
+            if local_user_id:
+                local_user_id_to_einfra_id[local_user_id] = einfra_id
+        if not local_user_id_to_einfra_id:
             continue
 
-        user = User.query.filter_by(id=local_user_id).one()
+        # bulk get users from the database
+        local_users = (
+            db.session.query(User)  # type: ignore
+            .filter(User.id.in_(local_user_id_to_einfra_id.keys()))
+            .all()
+        )
 
-        update_user_metadata(
-            user, aai_user.full_name, aai_user.email, aai_user.organization
+        # bulk get communities for the users
+        local_community_roles_by_user_id = (
+            community_support.get_user_list_community_membership(
+                local_user_id_to_einfra_id.keys()
+            )
         )
-        community_support.set_user_community_membership(user, aai_user.roles)
+
+        for user in local_users:
+            aai_user = aai_user_chunk_by_einfra_id[local_user_id_to_einfra_id[user.id]]
+            log.info("Setting user %s with roles %s", user, aai_user.roles)
+            update_user_metadata(
+                user, aai_user.full_name, aai_user.email, aai_user.organization
+            )
+
+            new_community_roles = filter_community_roles(
+                community_support, aai_user.roles
+            )
+
+            community_support.set_user_community_membership(
+                user,
+                new_community_roles=new_community_roles,
+                current_community_roles=local_community_roles_by_user_id.get(
+                    user.id, set()
+                ),
+            )
+
+        for unknown_user in set(aai_user_chunk_by_einfra_id.keys()) - set(
+            local_user_id_to_einfra_id.values()
+        ):
+            log.info(
+                "User with einfra id %s not yet found in the local database",
+                unknown_user,
+            )
+
     # for users that are not in the dump anymore, remove all communities
     for local_user_id in local_users_by_einfra.values():
         user = User.query.filter_by(id=local_user_id).one()
+        log.info("Removing obsolete user %s", user)
         community_support.set_user_community_membership(user, set())
```

Comment:
Should we also disable the account if possible? If not in dump, the user has been removed from all repository resources in Perun and shouldn't probably have access to any authenticated resources, not just communities?

Reply:
For the invitation to work, the VO needs to be currently set to auto enrollment - this is perun limitation. Thus any user in theory can log in to perun, join the VO which will be automatically approved and then can log in to the repository (will not have any communitied, but can log in). This means that repository must not differentiate between AnyUser and AuthenticatedUser (permissions must be based on roles & community roles only)